File: schemas/podcast-platform/drizzle/saved_episodes.ts
// Episodes bookmarked by users for later listening
// See README.md for full design rationale.

import { pgTable, uuid, timestamp, unique, index } from "drizzle-orm/pg-core";
import { users } from "./users";
import { episodes } from "./episodes";

export const savedEpisodes = pgTable(
  "saved_episodes",
  {
    id: uuid("id").primaryKey().defaultRandom(),
    userId: uuid("user_id").notNull().references(() => users.id, { onDelete: "cascade" }),
    episodeId: uuid("episode_id").notNull().references(() => episodes.id, { onDelete: "cascade" }),
    createdAt: timestamp("created_at", { withTimezone: true }).notNull().defaultNow(),
  },
  (table) => [
    unique("uq_saved_episodes_user_id_episode_id").on(table.userId, table.episodeId),
    index("idx_saved_episodes_user_id_created_at").on(table.userId, table.createdAt),
  ]
);

File: schemas/podcast-platform/spacetimedb/saved_episodes.rs
// saved_episodes: Episodes bookmarked by users for later listening.
// See README.md for full design rationale.

use spacetimedb::Timestamp;

#[spacetimedb::table(name = saved_episodes, public)]
pub struct SavedEpisode {
    #[primary_key]
    pub id: String, // UUID
    #[index(btree)]
    pub user_id: String, // UUID — FK → users.id (cascade delete)
    #[index(btree)]
    pub episode_id: String, // UUID — FK → episodes.id (cascade delete)
    pub created_at: Timestamp,
}
// Composite unique: (user_id, episode_id) — enforce in reducer logic
// Composite index: (user_id, created_at) — not supported, enforce in reducer logic
